x/oracle/genesis.go

```diff
@@ -0,0 +1,134 @@
+package oracle
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/persistenceOne/persistence-sdk/x/oracle/keeper"
+	"github.com/persistenceOne/persistence-sdk/x/oracle/types"
+)
+
+// InitGenesis initializes the x/oracle module's state from a provided genesis
+// state.
+func InitGenesis(ctx sdk.Context, keeper keeper.Keeper, genState types.GenesisState) {
+	for _, d := range genState.FeederDelegations {
+		voter, err := sdk.ValAddressFromBech32(d.ValidatorAddress)
+		if err != nil {
+			panic(err)
+		}
+
+		feeder, err := sdk.AccAddressFromBech32(d.FeederAddress)
+		if err != nil {
+			panic(err)
+		}
+
+		keeper.SetFeederDelegation(ctx, voter, feeder)
+	}
+
+	for _, ex := range genState.ExchangeRates {
+		keeper.SetExchangeRate(ctx, ex.Denom, ex.ExchangeRate)
+	}
+
+	for _, mc := range genState.MissCounters {
+		operator, err := sdk.ValAddressFromBech32(mc.ValidatorAddress)
+		if err != nil {
+			panic(err)
+		}
+
+		keeper.SetMissCounter(ctx, operator, mc.MissCounter)
+	}
+
+	for _, ap := range genState.AggregateExchangeRatePrevotes {
+		valAddr, err := sdk.ValAddressFromBech32(ap.Voter)
+		if err != nil {
+			panic(err)
+		}
+
+		keeper.SetAggregateExchangeRatePrevote(ctx, valAddr, ap)
+	}
+
+	for _, av := range genState.AggregateExchangeRateVotes {
+		valAddr, err := sdk.ValAddressFromBech32(av.Voter)
+		if err != nil {
+			panic(err)
+		}
+
+		keeper.SetAggregateExchangeRateVote(ctx, valAddr, av)
+	}
+
+	keeper.SetParams(ctx, genState.Params)
+
+	// check if the module account exists
+	moduleAcc := keeper.GetOracleAccount(ctx)
+	if moduleAcc == nil {
+		panic(fmt.Sprintf("%s module account has not been set", types.ModuleName))
+	}
+}
+
+// ExportGenesis returns the x/oracle module's exported genesis.
+func ExportGenesis(ctx sdk.Context, keeper keeper.Keeper) *types.GenesisState {
```

Comment:
Added the unit test.
Can you elaborate on the second part?

Reply:
@Tikaryan 

Go test has a flag:

```
-run regexp
    Run only those tests and examples matching the regular
    expression.
```

Since we use testify package, the flag is a bit trickier: `-testify.m`

Example of running `x/oracle/keeper` test suite with cover profile of genesis import/export only:
```
cd persistence-sdk/x/oracle/

go test ./keeper -coverprofile profile.out -testify.m TestOracleInitGenesis
go tool cover -html=profile.out
rm profile.out
```

It allows to see which code Genesis Import/export touched, especially this is super important to check that all state has been exported and imported during the test, so the code that actually iterates the state must be fully covered.

**Note:** the example works only on my branch, since I moved genesis logic into keeper, the coverage won't display anything for parent source dir. In your branch `genesis.go` is on the upper level, but the test itself is inside keeper.

Reply:
<img width="1132" alt="Screenshot 2023-02-24 at 01 57 21" src="https://user-images.githubusercontent.com/477998/221066331-4f5ff4b6-ecb9-4143-be27-004d79bb892d.png">

Reply:
P.S. panics are not necessarily to test there, I meant the "useful" part of the code should be covered.